### PR TITLE
Last obs before exposure flag and baseline flag

### PR DESCRIPTION
**Pull request by kamilsi**
@rammprasad, please verify the functionality of the new code additions. The existing examples confirm basic operations, but I was unable to fully test the baseline visit logic due to unmarked baseline visits in the study configuration file. Your insights or additional test cases would be helpful.

### CODE REVIEW

**Comment by ramiromagno** [comment]
Hi @rammprasad: @kamilsi asked for my review too but I noticed you pushed a few commits just an hour ago. Do you intend to make further changes until tomorrow's meeting or may look into it?

**Comment by rammprasad** [comment]
> Hi @rammprasad: @kamilsi asked for my review too but I noticed you pushed a few commits just an hour ago. Do you intend to make further changes until tomorrow's meeting or may look into it?

This is not ready for review. I have requested Kamil to look at my changes, replace the assertions, and add a test case. We will try and send it out for review tomorrow.

**Review by rammprasad — DISMISSED** [review]
Working as expected. REady for second review.

**Comment by rammprasad** [comment]
@kamilsi - this is working as expected. I made minor tweaks. Can you resolve the merge conflicts and prepare it for a second review and merge?

File: R/derive_blfl.R
Comment (ramiromagno) [inline R/derive_blfl.R:505]: no need to even have the prefix `sdtm.oak::` for functions of the same package.

File: R/derive_blfl.R
Comment (ramiromagno) [inline R/derive_blfl.R:178]: Suggestion: we should probably move these data set examples to inst/domain and made them available with `read_domain_example("vs")` thus simplifying the code at the Examples section.

Comment (rammprasad) [reply R/derive_blfl.R:178]: I thought about it, I am not sure if we will use this in another example.

**Comment by rammprasad** [comment]
@ShiyuC @ramiromagno - Ready for your review.

File: tests/testthat/test-derive_blfl.R
Comment (ShiyuC) [inline tests/testthat/test-derive_blfl.R:37]: Do we need to print observed_output here?

File: R/derive_blfl.R
Comment (ShiyuC) [inline R/derive_blfl.R:359]:
```suggestion
  # Assert that the input dataset has required columns
```

File: R/derive_blfl.R
Comment (ShiyuC) [inline R/derive_blfl.R:318]: Shall we add tests/examples of how "baseline_timepoints" is used?

File: tests/testthat/test-derive_blfl.R
Comment (rammprasad) [reply tests/testthat/test-derive_blfl.R:53]: I think the package version you have in your session may have some differences. The tests pass in the pipeline, and they also pass for me.

File: R/derive_blfl.R
Comment (ramiromagno) [reply R/derive_blfl.R:178]: Even if we don't, for readability sake it would be very beneficial IMO.

File: man/derive_blfl.Rd
Comment (ramiromagno) [inline man/derive_blfl.Rd:1]: Four comments:
  
 - parsing is way too relaxed for date-time components, so`"8931928-1231231-123123"` is parsed successfully. IMO it should result in `NA`.
 - Returned ISO 8601 times are not strictly conforming to the ISO 8601 format, e.g. `"12"` should be "-----T12".
 - Not sure if on empty strings (`""`) we want to return an empty string or `NA`...? I think in other date time related
 functions we decided to go with `NA`.
 - It is not totally clear to me what `ignore_seconds` should mean. It can either mean ignore on the input or ignore it on the output... And its interaction with `partial_as_na` should be clarified too. So perhaps this can be improved in the documentation.
 
And we can advantage of existing functions (namely `parse_dttm()` and `format_iso8601()`) for the parsing and that solve all the issues I mentioned above:

```r
dtc_datepart2 <- function(dtc, partial_as_na = TRUE) {
 
  date_fmts <- c("y-m-dTH:M:S", "y-m-dTH:M", "y-m-dTH", "y-m-dT", "y-m-d", "y-m", "y")
  fmt <- if (partial_as_na) date_fmts[1L:5L] else date_fmts
  
  dtc |>
    parse_dttm(fmt = fmt) |>
    magrittr::inset(, c("hour", "min", "sec"), NA_character_) |>
    format_iso8601()
}

dtc_timepart2 <- function(dtc, partial_as_na = TRUE, ignore_seconds = TRUE) {

  time_fmts <- c("y-m-dTH:M:S", "y-m-dTH:M", "y-m-dTH")
  
  if (isTRUE(partial_as_na) && isTRUE(ignore_seconds) ) fmt <- time_fmts[1L:2L]
  if (isTRUE(partial_as_na) && isFALSE(ignore_seconds)) fmt <- time_fmts[1L]
  if (isFALSE(partial_as_na)) fmt <- time_fmts
  
  ignore <- c("year", "mon", "mday", `if`(ignore_seconds, "sec", NULL))
  
  dtc |>
    parse_dttm(fmt = fmt) |>
    magrittr::inset(, ignore, NA_character_) |>
    format_iso8601()
}
```

File: R/derive_blfl.R
Comment (ramiromagno) [inline R/derive_blfl.R:1]: I have just started out reading `derive_blfl()` code and there are a few questions I have:

- Why create a look up table `domain_prefixed_names()` between variable names without the domain prefix and with it? This implies that throughout the code we have expressions like `domain_prefixed_names["orres"]` where I believe `"VSORRES"` would be simpler and more readable.
- Why assert on `sdtm_in` in two separate moments. You check first for the variables `c("DOMAIN", oak_id_vars())` and then also for "VSORRES", "VSSTAT", "VSTESTCD" and "VSDTC". Why not do it in one go right at the beginning?
- Why keep overwriting the object `ds_mod`? This just makes debugging a bit more difficult. I would rather have differently named variables for the different stages of the function computation, and perhaps encapsulate some of the major steps into their own functions.
- Filtering of rows with `NA` (or other values to be understood as `NA`, e.g. `"ND"` or `"NOT FOUND"`) would benefit from having a dedicated function similar to `tidyr::drop_na()` but that would support a parameter for handling these special values to be regarded as `NA`. Then the code:
   ```r
  bad_orres_rows <- is.na(ds_mod[[domain_prefixed_names["orres"]]]) |
  trimws(ds_mod[[domain_prefixed_names["orres"]]]) %in% c("ND", "NOT DONE", "")
  ds_mod <- ds_mod[!bad_orres_rows, ]
   ```
   could be replaced by something like:

  ```r
  sdtm_drop_na(ds_mod, .na = c("ND", "NOT DONE", ""))
  ```
- Filtering of rows where `--STAT` is not equal to `"NOT DONE"` could be simply rewritten as:
  ```r
  dplyr::filter(ds_mod, VSSTAT != "NOT DONE")
  ```

- Does the value `"<unspecified>"` as used when `VISIT` or `VSTPT` is missing has a special meaning in SDTM? From what I could find in the standard only the variable `RSCAT` may get a value similar to that one, and that is `"UNSPECIFIED"`. So if the value `"<unspecified>"` does not have a special SDTM meaning why not just use `NA_character_`?
- I'm very much confused by the use of `get()` in:
  ```r
    ds_subset_eq_2 <-
    ds_subset |>
    dplyr::filter(
      dom_dt == ref_dt,
      is.na(dom_tm) | is.na(ref_tm) | dom_tm == ref_tm,
      (VISIT %in% baseline_visits & get(domain_prefixed_names["tpt"]) %in% baseline_timepoints) |
        (VISIT %in% baseline_visits & length(baseline_timepoints) == 0L) |
        (get(domain_prefixed_names["tpt"]) %in% baseline_timepoints & length(baseline_visits) == 0L)
    )
  ```
- If you're using dplyr verbs throughout most of the code why not use `dplyr::bind_rows()` instead of `rbind()` in:
  ```r
  # Combine (*A) and (*B) and (*C)
  ds_base <- rbind(ds_subset_lt, ds_subset_eq_1, ds_subset_eq_2)
  ```
- `dplyr::arrange_at()` has been superseded in more recent versions of dplyr, doesn't `dplyr::arrange()` suffice in this case?
  ```r
  # Sort the rows in ascending order with respect to columns from con_col
  ds_base <- dplyr::arrange_at(ds_base, c("USUBJID", con_col))
  ```
- Can't we use the same terminology as in other derivation functions for the major input parameters, like `raw_dat` and `tgt_dat`, instead of `sdtm_in` and `dm_domain`, or are these fundamentally different?
- Is it important for `derive_blfl()` to support conditioned data frames?

Comment (ramiromagno) [reply R/derive_blfl.R:1]: I'm guessing that having to indicate SDTM variables will be very common, would it not be beneficial to have a very short named function that would return the full SDTM name of variable in the context of a domain? I'm thinking that it would be useful to replace locally defined look-up tables such as `domain_prefixed_names`. As an example, a `v()` function could consult package-level defined look up tables instead and return the appropriate variable names.

Example:

```r
v("vs", "stat") # "VSSTAT"
v("vs", "domain") # "DOMAIN"
v("vs", c("stat", "testcd", "tpt")) # "VSSTAT" "VSTESTCD" "VSTPT"
```

**Comment by rammprasad** [comment]
I have fixed a few minor items. Once this PR is approved, we can address the issues below.

1. Refactor code based on suggestions - https://github.com/pharmaverse/sdtm.oak/pull/49#discussion_r1645492545 and https://github.com/pharmaverse/sdtm.oak/pull/49#discussion_r1645420117
2. Validate the ISO date formats.

File: R/derive_blfl.R
Comment (rammprasad) [reply R/derive_blfl.R:1]: Will be addressed in a new MR

**Comment by rammprasad** [comment]
@ShiyuC @ramiromagno - Ready for your approval

**Comment by ramiromagno** [comment]
I'm getting this check message about potential spelling errors:

```
   < Potential spelling errors:
   <   WORD        FOUND IN
   < RFSTDTC     derive_blfl.Rd:27
   < RFXSTDTC    derive_blfl.Rd:28
   < TPT         derive_blfl.Rd:35
   < pre         derive_blfl.Rd:77
   < timpoints   derive_blfl.Rd:35
   < varaible    derive_blfl.Rd:19
   < xxTPT       derive_blfl.Rd:73
   < If these are false positive, run `spelling::update_wordlist()`.All Done!
```

**Review by ShiyuC — APPROVED** [review]
Thanks Ram! I’m ooo this afternoon but I’m happy to give the approval now as I saw my comments are resolved. If another review is needed, I’ll come back this evening to do it.

**Comment by edgar-manukyan** [comment]
> I'm getting this check message about potential spelling errors:
> 
> ```
>    < Potential spelling errors:
>    <   WORD        FOUND IN
>    < RFSTDTC     derive_blfl.Rd:27
>    < RFXSTDTC    derive_blfl.Rd:28
>    < TPT         derive_blfl.Rd:35
>    < pre         derive_blfl.Rd:77
>    < timpoints   derive_blfl.Rd:35
>    < varaible    derive_blfl.Rd:19
>    < xxTPT       derive_blfl.Rd:73
>    < If these are false positive, run `spelling::update_wordlist()`.All Done!
> ```

I will address this.

**Comment by ramiromagno** [comment]
Edgar: I was getting that too earlier but no longer with the most recent
version of the code. Have you pulled meanwhile?

**Comment by edgar-manukyan** [comment]
@ramiromagno, correct, I pulled and I got even more spelling mistakes, will push shortly.

**Comment by edgar-manukyan** [comment]
@ramiromagno, sorry for confusion, I was fixing in the `51-documentation-updates-v01`

**Comment by edgar-manukyan** [comment]
Now I switched to `18-last-obs-before-exposure-flag-and-baseline-flag` and I see the following and will push shortly
```
> spelling::spell_check_package()
  WORD        FOUND IN
CMGRPID     cnd_df.Rmd:49
CMTRT       cnd_df.Rmd:49
eCRF        algorithms.Rmd:15
            study_sdtm_spec.Rmd:44,55,58,58,61
eDT         algorithms.Rmd:15
            study_sdtm_spec.Rmd:44,55,71,73
MDR         algorithms.Rmd:24
            study_sdtm_spec.Rmd:40
pre         derive_blfl.Rd:77
            algorithms.Rmd:24
px          algorithms.Rmd:34,211
RFSTDTC     derive_blfl.Rd:27
RFXSTDTC    derive_blfl.Rd:28
timpoints   derive_blfl.Rd:35
TPT         derive_blfl.Rd:35
varaible    derive_blfl.Rd:19
xxTPT       derive_blfl.Rd:73
```

**Comment by ramiromagno** [comment]
Have you pulled? :) because those spelling issues should have been fixed by
now...

**Comment by edgar-manukyan** [comment]
@ramiromagno, the spellings should be good now. Can you please approve?

**Comment by edgar-manukyan** [comment]
> Have you pulled? :) because those spelling issues should have been fixed by now...
 I pulled and there were few findings, I fixed those and added the words from the https://github.com/pharmaverse/sdtm.oak/pull/56 as well.

**Review by ramiromagno — APPROVED** [review]
All good.

**Comment by edgar-manukyan** [comment]
I will fix the confilicts

**Pull request merged**
